File: plugins/opencode/src/autotag.ts
// Resolve the client tags a session opts into with
// AGENTO11Y_AUTO_CODING_AGENT_TAGS.
//
// Two variables drive it. AGENTO11Y_AUTO_CODING_AGENT_TAGS is the on/off switch
// and is off by default. AGENTO11Y_AUTO_CODING_AGENT_TAGS_NAMES is an optional
// allowlist that narrows the switch to some of the supported names; with the
// switch on and no list, every name applies.
//
// Client tags are the only tag mechanism that becomes a metric label
// (docs/concepts/tags-and-metadata.md), which is why the resolved values are
// attached to the client rather than to each generation. Nothing resolves
// until the switch is on: with the switch off, `resolveAutoTags` runs no git
// commands and no account lookup.
//
// Mirrors plugins/agento11y/internal/autotag/autotag.go. The Go launcher owns
// the hook-based agents; this file covers the in-process opencode plugin, the
// same way tags.ts mirrors the Go built-in tag builder.

import { userInfo } from "node:os";
import { resolveGitBranch, resolveGitRepo } from "./git.js";

/** One value the plugin can resolve for the session. */
export type AutoTag = "user" | "repo" | "branch";

/** Supported names in the order diagnostics and tag maps use. */
export const AUTO_TAG_ORDER: readonly AutoTag[] = ["user", "repo", "branch"];

/** Shorthand that enables every supported name. */
export const AUTO_TAG_ALL = "all";

/**
 * Suffixes of the two branded variables that drive automatic tags:
 * AUTO_TAGS_SUFFIX is the on/off switch, AUTO_TAG_NAMES_SUFFIX the optional
 * allowlist.
 */
export const AUTO_TAGS_SUFFIX = "AUTO_CODING_AGENT_TAGS";
export const AUTO_TAG_NAMES_SUFFIX = "AUTO_CODING_AGENT_TAGS_NAMES";

/**
 * Cap on a resolved value, in Unicode code points. Every value becomes a
 * Prometheus label, so an accidentally long branch name or email stays
 * bounded. Truncation keeps the start of the value.
 */
export const MAX_AUTO_TAG_VALUE_LENGTH = 128;

/**
 * Tag key each name is written under. The branch reuses the per-generation
 * built-in key `git.branch` so the export carries one branch key: the
 * per-generation tag wins in the generation export, and this client tag
 * supplies the metric label.
 */
const TAG_KEYS: Record<AutoTag, string> = {
  user: "user",
  repo: "repo",
  branch: "git.branch",
};

/** The client tag key an auto-tag name is written under. */
export function autoTagKey(name: AutoTag): string {
  return TAG_KEYS[name];
}

export interface ParsedAutoTags {
  enabled: Set<AutoTag>;
  /** Names that are not supported, lowercased, for the caller to log. */
  unknown: string[];
}

/**
 * Parse AGENTO11Y_AUTO_CODING_AGENT_TAGS_NAMES ("user,repo") into the set of
 * enabled names. Matching is case-insensitive and surrounding whitespace is
 * trimmed. "all" enables every supported name, which is also what an unset
 * allowlist means. Empty entries are skipped; unrecognized names are returned
 * separately so a caller can log them and keep going with the names it did
 * recognize.
 */
export function parseAutoTags(raw: string): ParsedAutoTags {
  const enabled = new Set<AutoTag>();
  const unknown: string[] = [];
  if (raw.trim() === "") return { enabled, unknown };
  for (const field of raw.split(",")) {
    const name = field.trim().toLowerCase();
    if (name === "") continue;
    if (name === AUTO_TAG_ALL) {
      for (const supported of AUTO_TAG_ORDER) enabled.add(supported);
      continue;
    }
    if (!isAutoTag(name)) {
      if (!unknown.includes(name)) unknown.push(name);
      continue;
    }
    enabled.add(name);
  }
  return { enabled, unknown };
}

function isAutoTag(name: string): name is AutoTag {
  return (AUTO_TAG_ORDER as readonly string[]).includes(name);
}

/** One branded variable resolved to its value and the spelling that won. */
export type BrandedLookup = (
  suffix: string,
) => { value: string; key: string } | undefined;

export interface AutoTagSelection {
  /** The switch holds a true value. It can be true with an empty `enabled`, when the allowlist named nothing supported. */
  on: boolean;
  /** Names to resolve: every supported name, unless the allowlist narrows it. */
  enabled: Set<AutoTag>;
  /** Allowlist entries that name no supported value, lowercased. */
  unknown: string[];
  /** The allowlist variable held a value, which separates "narrowed to nothing" from "not narrowed". */
  namesSet: boolean;
}

/**
 * Read the switch and the allowlist and report which names to resolve. Mirrors
 * autotag.Select on the Go side.
 *
 * Every problem is reported through `warn` and none of them throws: a switch
 * value that is not a boolean leaves the mechanism off, an unsupported name is
 * skipped, and an allowlist set while the switch is off attaches nothing.
 */
export function selectAutoTags(
  lookup: BrandedLookup,
  warn: (message: string) => void,
): AutoTagSelection {
  const names = lookup(AUTO_TAG_NAMES_SUFFIX);
  const selection: AutoTagSelection = {
    on: false,
    enabled: new Set<AutoTag>(),
    unknown: [],
    namesSet: names !== undefined,
  };

  const configured = lookup(AUTO_TAGS_SUFFIX);
  if (configured !== undefined) {
    const on = parseAutoTagsSwitch(configured.value);
    if (on === undefined) {
      warn(
        `invalid ${configured.key}="${configured.value}": expected a boolean, and the names go in AGENTO11Y_${AUTO_TAG_NAMES_SUFFIX}`,
      );
    }
    selection.on = on ?? false;
  }
  if (!selection.on) {
    if (names !== undefined) {
      warn(
        `${names.key} is set but AGENTO11Y_${AUTO_TAGS_SUFFIX} is off, so no automatic tags are attached`,
      );
    }
    return selection;
  }
  if (names === undefined) {
    selection.enabled = new Set(AUTO_TAG_ORDER);
    return selection;
  }

  const parsed = parseAutoTags(names.value);
  selection.enabled = parsed.enabled;
  selection.unknown = parsed.unknown;
  if (parsed.unknown.length > 0) {
    warn(
      `${names.key} has unsupported names ${parsed.unknown.join(", ")}; supported: ${[...AUTO_TAG_ORDER, AUTO_TAG_ALL].join(", ")}`,
    );
  }
  if (parsed.enabled.size === 0) {
    warn(
      `${names.key} names no supported value, so no automatic tags are attached`,
    );
  }
  return selection;
}

// The same boolean whitelist config.ts applies to every other branded flag,
// kept here so the switch and its diagnostic live together.
function parseAutoTagsSwitch(raw: string): boolean | undefined {
  const normalized = raw.trim().toLowerCase();
  if (["1", "true", "yes", "on"].includes(normalized)) return true;
  if (["0", "false", "no", "off"].includes(normalized)) return false;
  return undefined;
}

/**
 * Parse a comma-separated "key=value" string (AGENTO11Y_TAGS) into a tag map.
 * Malformed entries (empty keys, missing '=', empty values) are skipped.
 * Mirrors envconfig.ParseExtraTags on the Go side.
 */
export function parseTagPairs(raw: string | undefined): Record<string, string> {
  const out: Record<string, string> = {};
  if (!raw || raw.trim() === "") return out;
  for (const pair of raw.split(",")) {
    const eq = pair.indexOf("=");
    if (eq === -1) continue;
    const key = pair.slice(0, eq).trim();
    const value = pair.slice(eq + 1).trim();
    if (key === "" || value === "") continue;
    out[key] = value;
  }
  return out;
}

export interface AutoTagInputs {
  /** Workspace root the repository and branch are resolved from. */
  cwd?: string;
  /**
   * Identity the caller already resolved, usually AGENTO11Y_USER_ID. It ranks
   * above the OS account name.
   */
  userId?: string;
  /**
   * Tags AGENTO11Y_TAGS defines. A key set there wins, so the resolver leaves
   * it out: the SDK merges caller tags over environment tags, which is the
   * opposite of the precedence this switch needs.
   */
  explicitTags?: Record<string, string>;
}

/**
 * Resolve the client tags for the enabled names. Values are trimmed and capped
 * at MAX_AUTO_TAG_VALUE_LENGTH; a name that resolves to nothing leaves its key
 * off, and so does a key AGENTO11Y_TAGS already defines. Returns undefined
 * when nothing resolves.
 */
export function resolveAutoTags(
  enabled: Set<AutoTag>,
  inputs: AutoTagInputs = {},
): Record<string, string> | undefined {
  if (enabled.size === 0) return undefined;
  const explicit = inputs.explicitTags ?? {};
  const tags: Record<string, string> = {};
  for (const name of AUTO_TAG_ORDER) {
    if (!enabled.has(name)) continue;
    const key = autoTagKey(name);
    if (Object.hasOwn(explicit, key)) continue;
    const value = clean(resolveOne(name, inputs));
    if (value === "") continue;
    tags[key] = value;
  }
  return Object.keys(tags).length > 0 ? tags : undefined;
}

function resolveOne(name: AutoTag, inputs: AutoTagInputs): string | undefined {
  switch (name) {
    case "user":
      return resolveUser(inputs);
    case "repo":
      return inputs.cwd ? resolveGitRepo(inputs.cwd) : undefined;
    case "branch":
      return inputs.cwd ? resolveGitBranch(inputs.cwd) : undefined;
  }
}

/**
 * Prefer the identity the user configured, then the OS account name. opencode
 * has no signed-in user of its own, so there is no middle source here; the Go
 * launcher inserts the agent-resolved identity between the two.
 */
function resolveUser(inputs: AutoTagInputs): string | undefined {
  const configured = (inputs.userId ?? "").trim();
  if (configured !== "") return configured;
  try {
    return userInfo().username;
  } catch {
    return undefined;
  }
}

function clean(value: string | undefined): string {
  const trimmed = (value ?? "").trim();
  const points = Array.from(trimmed);
  return points.length > MAX_AUTO_TAG_VALUE_LENGTH
    ? points.slice(0, MAX_AUTO_TAG_VALUE_LENGTH).join("")
    : trimmed;
}

File: plugins/agento11y/internal/autotag/autotag.go
// Package autotag resolves the client tags a session opts into with
// AGENTO11Y_AUTO_CODING_AGENT_TAGS. The launcher already knows who is running
// it, which repository the session is in, and which branch is checked out;
// this package turns the enabled names into client tags so those values reach
// OTel metrics.
//
// Two variables drive it. AGENTO11Y_AUTO_CODING_AGENT_TAGS is the on/off
// switch and is off by default. AGENTO11Y_AUTO_CODING_AGENT_TAGS_NAMES is an
// optional allowlist that narrows the switch to some of the supported names;
// with the switch on and no list, every name applies.
//
// Client tags are the only tag mechanism that becomes a metric label
// (docs/concepts/tags-and-metadata.md), which is why the resolved values are
// attached here rather than as per-generation tags. Nothing resolves until the
// switch is on: with the switch off, Resolve reads no files, runs no lookups,
// and returns nil.
package autotag

import (
	"log"
	"os"
	"os/user"
	"strings"

	"github.com/grafana/agento11y/plugins/agento11y/internal/envconfig"
	"github.com/grafana/agento11y/plugins/agento11y/internal/gitbranch"
)

// Tag keys the enabled names are written under. The branch reuses the
// per-generation built-in key `git.branch` so the export carries one branch
// key: the built-in wins in the generation export, and this client tag
// supplies the metric label.
const (
	KeyUser   = "user"
	KeyRepo   = "repo"
	KeyBranch = "git.branch"
)

// MaxValueLen caps a resolved value at 128 Unicode code points. Every value
// becomes a Prometheus label, so an accidentally long branch name or email
// stays bounded. Truncation keeps the start of the value.
const MaxValueLen = 128

// osUsername returns the login name of the account running the process.
// Indirected through a variable so tests can drive the last user fallback
// without depending on the machine they run on.
var osUsername = func() string {
	u, err := user.Current()
	if err != nil {
		return ""
	}
	return u.Username
}

// Inputs carry what the caller already knows. Everything else is resolved
// here.
type Inputs struct {
	// Cwd is the workspace root used for repository and branch resolution.
	// Empty falls back to the process working directory, which is the
	// directory the host agent started the hook in.
	Cwd string
	// UserID is the identity the agent adapter already resolved, if any. It
	// is the fallback below the configured USER_ID and above the OS account
	// name.
	UserID string
	// Lookup reads branded variables. nil uses the process environment;
	// doctor passes a lookup backed by its pre-merge snapshot so it reports
	// the values the hooks resolve.
	Lookup envconfig.Lookup
}

// Result is the full picture of one resolution, for callers that report on it
// rather than just consuming the tags.
type Result struct {
	// Tags are the client tags to pass to the SDK: resolved values, minus any
	// key an explicit AGENTO11Y_TAGS entry already defines. nil when empty.
	Tags map[string]string
	// Values is the resolved value per enabled name before explicit tags are
	// applied. A name that resolved to nothing is absent.
	Values map[envconfig.AutoTag]string
	// Shadowed lists enabled names whose key AGENTO11Y_TAGS already defines,
	// in AutoTagOrder. The explicit value wins, so these are not in Tags.
	Shadowed []envconfig.AutoTag
}

// TagKey returns the client tag key an auto-tag name is written under.
func TagKey(t envconfig.AutoTag) string {
	switch t {
	case envconfig.AutoTagUser:
		return KeyUser
	case envconfig.AutoTagRepo:
		return KeyRepo
	case envconfig.AutoTagBranch:
		return KeyBranch
	default:
		return ""
	}
}

// Selection is what the two AUTO_CODING_AGENT_TAGS variables asked for.
type Selection struct {
	// On reports that the switch holds a true value. It can be true with an
	// empty Enabled set, when the allowlist named nothing supported.
	On bool
	// Enabled is the set of names to resolve: every supported name, unless the
	// allowlist narrows it.
	Enabled map[envconfig.AutoTag]bool
	// Unknown lists allowlist entries that name no supported value, lowercased.
	Unknown []string
	// NamesSet reports that the allowlist variable held a value. It separates a
	// selection narrowed to nothing from one that was never narrowed.
	NamesSet bool
}

// Select reads the switch and the allowlist and reports which names to
// resolve. look reads branded variables; nil uses the process environment.
//
// Every problem is logged and none stops the caller: a switch value that is
// not a boolean leaves the mechanism off, an unsupported name is skipped, and
// an allowlist set while the switch is off attaches nothing. Hooks must not
// write to stderr, so pass the adapter logger (nil stays silent).
func Select(look envconfig.Lookup, logger *log.Logger) Selection {
	if look == nil {
		look = envconfig.LookupEnv
	}
	namesRaw, namesKey, namesSet := look(envconfig.AutoTagNamesSuffix)
	sel := Selection{NamesSet: namesSet}

	raw, key, ok := look(envconfig.AutoTagsSuffix)
	if ok {
		on, valid := envconfig.ParseBoolValue(raw)
		if !valid && logger != nil {
			logger.Printf("config: invalid %s=%q; expected a boolean, and the names go in %s",
				key, raw, envconfig.PreferredKey(envconfig.AutoTagNamesSuffix))
		}
		sel.On = on
	}
	if !sel.On {
		if namesSet && logger != nil {
			logger.Printf("config: %s is set but %s is off, so no automatic tags are attached",
				namesKey, envconfig.PreferredKey(envconfig.AutoTagsSuffix))
		}
		return sel
	}
	if !namesSet {
		sel.Enabled = envconfig.AllAutoTags()
		return sel
	}
	sel.Enabled, sel.Unknown = envconfig.ParseAutoTags(namesRaw)
	if logger != nil {
		if len(sel.Unknown) > 0 {
			logger.Printf("config: %s has unsupported names %s; supported: %s",
				namesKey, strings.Join(sel.Unknown, ", "), SupportedNames())
		}
		if len(sel.Enabled) == 0 {
			logger.Printf("config: %s names no supported value, so no automatic tags are attached", namesKey)
		}
	}
	return sel
}

// SupportedNames lists the values the allowlist accepts, for the message that
// reports a rejected one.
func SupportedNames() string {
	names := make([]string, 0, len(envconfig.AutoTagOrder)+1)
	for _, name := range envconfig.AutoTagOrder {
		names = append(names, string(name))
	}
	return strings.Join(append(names, envconfig.AutoTagAll), ", ")
}

// FromEnv is the one call a client-construction path makes: it reads the
// AUTO_CODING_AGENT_TAGS family, resolves whatever it enables, and returns the
// client tags to hand to the SDK. It returns nil when the switch is off, so a
// session that has not opted in carries exactly the tags it carried before.
//
// Hooks must not write to stderr, so pass the adapter logger (nil stays
// silent); Select documents what gets logged.
//
// Replayed sessions must not take these values: the importer runs on a
// different day, possibly in a different checkout, so the current user,
// repository and branch do not describe the session it is importing. That is
// why internal/history/export.go does not call this.
func FromEnv(in Inputs, logger *log.Logger) map[string]string {
	if in.Lookup == nil {
		in.Lookup = envconfig.LookupEnv
	}
	sel := Select(in.Lookup, logger)
	if len(sel.Enabled) == 0 {
		return nil
	}
	return Resolve(sel.Enabled, in)
}

// Resolve returns the client tags for the enabled names. Values are trimmed
// and capped at MaxValueLen; a name that resolves to nothing leaves its key
// off. A key already set in AGENTO11Y_TAGS is left off too, because the SDK
// merges caller tags over environment tags and an explicit tag must win.
// Returns nil when nothing resolves.
func Resolve(enabled map[envconfig.AutoTag]bool, in Inputs) map[string]string {
	return Describe(enabled, in).Tags
}

// Describe resolves the enabled names and reports what each one produced,
// including the names an explicit tag shadows. `agento11y doctor` uses it to
// show the exact strings before they leave the machine; hooks use Resolve.
func Describe(enabled map[envconfig.AutoTag]bool, in Inputs) Result {
	if len(enabled) == 0 {
		return Result{}
	}
	look := in.Lookup
	if look == nil {
		look = envconfig.LookupEnv
	}
	if in.Cwd == "" {
		if wd, err := os.Getwd(); err == nil {
			in.Cwd = wd
		}
	}

	res := Result{Values: make(map[envconfig.AutoTag]string, len(enabled))}
	for _, name := range envconfig.AutoTagOrder {
		if !enabled[name] {
			continue
		}
		if v := clean(resolveOne(name, in, look)); v != "" {
			res.Values[name] = v
		}
	}
	if len(res.Values) == 0 {
		return Result{}
	}

	tagsValue, _, _ := look("TAGS")
	explicit := envconfig.ParseExtraTags(tagsValue)
	tags := make(map[string]string, len(res.Values))
	for _, name := range envconfig.AutoTagOrder {
		value, ok := res.Values[name]
		if !ok {
			continue
		}
		key := TagKey(name)
		if _, taken := explicit[key]; taken {
			res.Shadowed = append(res.Shadowed, name)
			continue
		}
		tags[key] = value
	}
	if len(tags) > 0 {
		res.Tags = tags
	}
	return res
}

func resolveOne(name envconfig.AutoTag, in Inputs, look envconfig.Lookup) string {
	switch name {
	case envconfig.AutoTagUser:
		return resolveUser(in, look)
	case envconfig.AutoTagRepo:
		return gitbranch.Repo(in.Cwd)
	case envconfig.AutoTagBranch:
		return gitbranch.Resolve(in.Cwd)
	default:
		return ""
	}
}

// resolveUser prefers the identity the user configured, then the one the agent
// adapter resolved from its own session data, then the OS account name.
func resolveUser(in Inputs, look envconfig.Lookup) string {
	if v, _, ok := look("USER_ID"); ok {
		return v
	}
	if v := strings.TrimSpace(in.UserID); v != "" {
		return v
	}
	return osUsername()
}

func clean(v string) string {
	v = strings.TrimSpace(v)
	runes := []rune(v)
	if len(runes) > MaxValueLen {
		return string(runes[:MaxValueLen])
	}
	return v
}
